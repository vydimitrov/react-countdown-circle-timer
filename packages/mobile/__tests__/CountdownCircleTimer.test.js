import React from 'react'
import { Text } from 'react-native'
import renderer from 'react-test-renderer'
import { render, waitFor } from '@testing-library/react-native'

import { CountdownCircleTimer } from '../src'

Math.random = () => 0.124578

const fixture = {
  duration: 10,
  colors: [
    ['#004777', 0.33],
    ['#F7B801', 0.33],
    ['#A30000', 0.33],
  ],
}

describe('snapshot tests', () => {
  it('renders', () => {
    const tree = renderer.create(<CountdownCircleTimer {...fixture} />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders with time', () => {
    const tree = renderer
      .create(
        <CountdownCircleTimer
          {...fixture}
          renderAriaTime={({ remainingTime }) => remainingTime}
        >
          {({ remainingTime }) => <Text>{remainingTime}</Text>}
        </CountdownCircleTimer>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders with single color', () => {
    const tree = renderer
      .create(
        <CountdownCircleTimer duration={5} colors={[['#004777', 1]]}>
          {({ remainingTime }) => <Text>{remainingTime}</Text>}
        </CountdownCircleTimer>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders with single color provided as a string', () => {
    const tree = renderer
      .create(
        <CountdownCircleTimer duration={5} colors="#004777">
          {({ remainingTime }) => <Text>{remainingTime}</Text>}
        </CountdownCircleTimer>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders with gradient', () => {
    const tree = renderer
      .create(
        <CountdownCircleTimer
          duration={5}
          colors={[
            ['#004777', 0.4],
            ['#aabbcc', 0.6],
          ]}
          isLinearGradient
        >
          {({ remainingTime }) => <Text>{remainingTime}</Text>}
        </CountdownCircleTimer>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders with different trail stroke width', () => {
    const tree = renderer
      .create(
        <CountdownCircleTimer
          {...fixture}
          trailStrokeWidth={16}
          strokeWidth={14}
        >
          {({ remainingTime }) => <Text>{remainingTime}</Text>}
        </CountdownCircleTimer>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders correctly when color duration is not provided', () => {
    const tree = renderer
      .create(
        <CountdownCircleTimer
          {...fixture}
          colors={[['#047'], ['#aaa'], ['#bbb']]}
        >
          {({ remainingTime }) => <Text>{remainingTime}</Text>}
        </CountdownCircleTimer>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})

describe('functional tests', () => {
  it('should start the timer from value provided in initialRemainingTime', () => {
    const { getByText } = render(
      <CountdownCircleTimer {...fixture} initialRemainingTime={3.7}>
        {({ remainingTime }) => <Text>{remainingTime}</Text>}
      </CountdownCircleTimer>
    )

    expect(getByText('4')).toBeTruthy()
  })
})

describe('behaviour tests', () => {
  it('should call onComplete at the end of the countdown', async () => {
    const onComplete = jest.fn()
    const { findByText } = render(
      <CountdownCircleTimer
        {...fixture}
        duration={1}
        isPlaying
        onComplete={onComplete}
      >
        {({ remainingTime }) => <Text>{remainingTime}</Text>}
      </CountdownCircleTimer>
    )

    expect(await findByText('0')).toBeTruthy()
    expect(onComplete).toHaveBeenCalledWith(1)
  })

  it('should clear repeat timeout when the component is unmounted', () => {
    const clearTimeoutMock = jest.fn()

    global.clearTimeout = clearTimeoutMock

    const { unmount } = render(<CountdownCircleTimer {...fixture} />)

    unmount()

    expect(clearTimeoutMock).toHaveBeenCalled()
  })
})
