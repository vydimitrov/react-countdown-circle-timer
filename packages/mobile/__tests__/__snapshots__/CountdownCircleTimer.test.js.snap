// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot tests renders 1`] = `
<View
  accessibilityLabel="Countdown timer"
  accessible={true}
  style={
    Object {
      "height": 180,
      "position": "relative",
      "width": 180,
    }
  }
>
  <RNSVGSvgView
    bbHeight={180}
    bbWidth={180}
    focusable={false}
    height={180}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {
          "flex": 0,
          "height": 180,
          "width": 180,
        },
      ]
    }
    width={180}
  >
    <RNSVGGroup>
      <RNSVGPath
        d="m 90,6
          a 84,84 0 1,0 0,168
          a 84,84 0 1,0 0,-168"
        fill={null}
        propList={
          Array [
            "fill",
            "stroke",
            "strokeWidth",
          ]
        }
        stroke={4292467161}
        strokeWidth={12}
      />
      <RNSVGPath
        d="m 90,6
          a 84,84 0 1,0 0,168
          a 84,84 0 1,0 0,-168"
        fill={null}
        propList={
          Array [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeDasharray",
            "strokeDashoffset",
            "strokeLinecap",
          ]
        }
        stroke={4278208375}
        strokeDasharray={
          Array [
            527.7875658030853,
            527.7875658030853,
          ]
        }
        strokeDashoffset={null}
        strokeLinecap={1}
        strokeWidth={12}
      />
    </RNSVGGroup>
  </RNSVGSvgView>
</View>
`;

exports[`snapshot tests renders correctly when color duration is not provided 1`] = `
<View
  accessibilityLabel="Countdown timer"
  accessible={true}
  style={
    Object {
      "height": 180,
      "position": "relative",
      "width": 180,
    }
  }
>
  <RNSVGSvgView
    bbHeight={180}
    bbWidth={180}
    focusable={false}
    height={180}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {
          "flex": 0,
          "height": 180,
          "width": 180,
        },
      ]
    }
    width={180}
  >
    <RNSVGGroup>
      <RNSVGPath
        d="m 90,6
          a 84,84 0 1,0 0,168
          a 84,84 0 1,0 0,-168"
        fill={null}
        propList={
          Array [
            "fill",
            "stroke",
            "strokeWidth",
          ]
        }
        stroke={4292467161}
        strokeWidth={12}
      />
      <RNSVGPath
        d="m 90,6
          a 84,84 0 1,0 0,168
          a 84,84 0 1,0 0,-168"
        fill={null}
        propList={
          Array [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeDasharray",
            "strokeDashoffset",
            "strokeLinecap",
          ]
        }
        stroke={4278207607}
        strokeDasharray={
          Array [
            527.7875658030853,
            527.7875658030853,
          ]
        }
        strokeDashoffset={null}
        strokeLinecap={1}
        strokeWidth={12}
      />
    </RNSVGGroup>
  </RNSVGSvgView>
  <View
    accessibilityElementsHidden={true}
    importantForAccessibility="no-hide-descendants"
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "height": "100%",
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": "100%",
      }
    }
  >
    <Text>
      10
    </Text>
  </View>
</View>
`;

exports[`snapshot tests renders with different trail stroke width 1`] = `
<View
  accessibilityLabel="Countdown timer"
  accessible={true}
  style={
    Object {
      "height": 180,
      "position": "relative",
      "width": 180,
    }
  }
>
  <RNSVGSvgView
    bbHeight={180}
    bbWidth={180}
    focusable={false}
    height={180}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {
          "flex": 0,
          "height": 180,
          "width": 180,
        },
      ]
    }
    width={180}
  >
    <RNSVGGroup>
      <RNSVGPath
        d="m 90,8
          a 82,82 0 1,0 0,164
          a 82,82 0 1,0 0,-164"
        fill={null}
        propList={
          Array [
            "fill",
            "stroke",
            "strokeWidth",
          ]
        }
        stroke={4292467161}
        strokeWidth={16}
      />
      <RNSVGPath
        d="m 90,8
          a 82,82 0 1,0 0,164
          a 82,82 0 1,0 0,-164"
        fill={null}
        propList={
          Array [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeDasharray",
            "strokeDashoffset",
            "strokeLinecap",
          ]
        }
        stroke={4278208375}
        strokeDasharray={
          Array [
            515.221195188726,
            515.221195188726,
          ]
        }
        strokeDashoffset={null}
        strokeLinecap={1}
        strokeWidth={14}
      />
    </RNSVGGroup>
  </RNSVGSvgView>
  <View
    accessibilityElementsHidden={true}
    importantForAccessibility="no-hide-descendants"
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "height": "100%",
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": "100%",
      }
    }
  >
    <Text>
      10
    </Text>
  </View>
</View>
`;

exports[`snapshot tests renders with gradient 1`] = `
<View
  accessibilityLabel="Countdown timer"
  accessible={true}
  style={
    Object {
      "height": 180,
      "position": "relative",
      "width": 180,
    }
  }
>
  <RNSVGSvgView
    bbHeight={180}
    bbWidth={180}
    focusable={false}
    height={180}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {
          "flex": 0,
          "height": 180,
          "width": 180,
        },
      ]
    }
    width={180}
  >
    <RNSVGGroup>
      <RNSVGDefs>
        <RNSVGLinearGradient
          gradient={
            Array [
              0,
              -16758921,
              0.4,
              -16758921,
              1,
              -5588020,
            ]
          }
          gradientTransform={null}
          gradientUnits={0}
          name="countdown-circle-timer-gradient-4hgb79ur1xg"
          x1="1"
          x2="0"
          y1="0"
          y2="0"
        />
      </RNSVGDefs>
      <RNSVGPath
        d="m 90,6
          a 84,84 0 1,0 0,168
          a 84,84 0 1,0 0,-168"
        fill={null}
        propList={
          Array [
            "fill",
            "stroke",
            "strokeWidth",
          ]
        }
        stroke={4292467161}
        strokeWidth={12}
      />
      <RNSVGPath
        d="m 90,6
          a 84,84 0 1,0 0,168
          a 84,84 0 1,0 0,-168"
        fill={null}
        propList={
          Array [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeDasharray",
            "strokeDashoffset",
            "strokeLinecap",
          ]
        }
        stroke={
          Array [
            1,
            "countdown-circle-timer-gradient-4hgb79ur1xg",
          ]
        }
        strokeDasharray={
          Array [
            527.7875658030853,
            527.7875658030853,
          ]
        }
        strokeDashoffset={null}
        strokeLinecap={1}
        strokeWidth={12}
      />
    </RNSVGGroup>
  </RNSVGSvgView>
  <View
    accessibilityElementsHidden={true}
    importantForAccessibility="no-hide-descendants"
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "height": "100%",
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": "100%",
      }
    }
  >
    <Text>
      5
    </Text>
  </View>
</View>
`;

exports[`snapshot tests renders with single color 1`] = `
<View
  accessibilityLabel="Countdown timer"
  accessible={true}
  style={
    Object {
      "height": 180,
      "position": "relative",
      "width": 180,
    }
  }
>
  <RNSVGSvgView
    bbHeight={180}
    bbWidth={180}
    focusable={false}
    height={180}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {
          "flex": 0,
          "height": 180,
          "width": 180,
        },
      ]
    }
    width={180}
  >
    <RNSVGGroup>
      <RNSVGPath
        d="m 90,6
          a 84,84 0 1,0 0,168
          a 84,84 0 1,0 0,-168"
        fill={null}
        propList={
          Array [
            "fill",
            "stroke",
            "strokeWidth",
          ]
        }
        stroke={4292467161}
        strokeWidth={12}
      />
      <RNSVGPath
        d="m 90,6
          a 84,84 0 1,0 0,168
          a 84,84 0 1,0 0,-168"
        fill={null}
        propList={
          Array [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeDasharray",
            "strokeDashoffset",
            "strokeLinecap",
          ]
        }
        stroke={4278208375}
        strokeDasharray={
          Array [
            527.7875658030853,
            527.7875658030853,
          ]
        }
        strokeDashoffset={null}
        strokeLinecap={1}
        strokeWidth={12}
      />
    </RNSVGGroup>
  </RNSVGSvgView>
  <View
    accessibilityElementsHidden={true}
    importantForAccessibility="no-hide-descendants"
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "height": "100%",
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": "100%",
      }
    }
  >
    <Text>
      5
    </Text>
  </View>
</View>
`;

exports[`snapshot tests renders with single color provided as a string 1`] = `
<View
  accessibilityLabel="Countdown timer"
  accessible={true}
  style={
    Object {
      "height": 180,
      "position": "relative",
      "width": 180,
    }
  }
>
  <RNSVGSvgView
    bbHeight={180}
    bbWidth={180}
    focusable={false}
    height={180}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {
          "flex": 0,
          "height": 180,
          "width": 180,
        },
      ]
    }
    width={180}
  >
    <RNSVGGroup>
      <RNSVGPath
        d="m 90,6
          a 84,84 0 1,0 0,168
          a 84,84 0 1,0 0,-168"
        fill={null}
        propList={
          Array [
            "fill",
            "stroke",
            "strokeWidth",
          ]
        }
        stroke={4292467161}
        strokeWidth={12}
      />
      <RNSVGPath
        d="m 90,6
          a 84,84 0 1,0 0,168
          a 84,84 0 1,0 0,-168"
        fill={null}
        propList={
          Array [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeDasharray",
            "strokeDashoffset",
            "strokeLinecap",
          ]
        }
        stroke={4278208375}
        strokeDasharray={
          Array [
            527.7875658030853,
            527.7875658030853,
          ]
        }
        strokeDashoffset={null}
        strokeLinecap={1}
        strokeWidth={12}
      />
    </RNSVGGroup>
  </RNSVGSvgView>
  <View
    accessibilityElementsHidden={true}
    importantForAccessibility="no-hide-descendants"
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "height": "100%",
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": "100%",
      }
    }
  >
    <Text>
      5
    </Text>
  </View>
</View>
`;

exports[`snapshot tests renders with time 1`] = `
<View
  accessibilityLabel="Countdown timer"
  accessible={true}
  style={
    Object {
      "height": 180,
      "position": "relative",
      "width": 180,
    }
  }
>
  <RNSVGSvgView
    bbHeight={180}
    bbWidth={180}
    focusable={false}
    height={180}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {
          "flex": 0,
          "height": 180,
          "width": 180,
        },
      ]
    }
    width={180}
  >
    <RNSVGGroup>
      <RNSVGPath
        d="m 90,6
          a 84,84 0 1,0 0,168
          a 84,84 0 1,0 0,-168"
        fill={null}
        propList={
          Array [
            "fill",
            "stroke",
            "strokeWidth",
          ]
        }
        stroke={4292467161}
        strokeWidth={12}
      />
      <RNSVGPath
        d="m 90,6
          a 84,84 0 1,0 0,168
          a 84,84 0 1,0 0,-168"
        fill={null}
        propList={
          Array [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeDasharray",
            "strokeDashoffset",
            "strokeLinecap",
          ]
        }
        stroke={4278208375}
        strokeDasharray={
          Array [
            527.7875658030853,
            527.7875658030853,
          ]
        }
        strokeDashoffset={null}
        strokeLinecap={1}
        strokeWidth={12}
      />
    </RNSVGGroup>
  </RNSVGSvgView>
  <View
    accessibilityElementsHidden={true}
    importantForAccessibility="no-hide-descendants"
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "height": "100%",
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": "100%",
      }
    }
  >
    <Text>
      10
    </Text>
  </View>
  <Text
    accessibilityLiveRegion="assertive"
    accessibilityRole="timer"
    importantForAccessibility="yes"
    style={
      Object {
        "height": 0,
        "opacity": 0,
      }
    }
  >
    10
  </Text>
</View>
`;
